### PR TITLE
Omar/hns sync

### PR DESCRIPTION
**Pull request by OmarKhatib158**
Added support for blobfs.
Currently it does not delete folders/files in the target

### CODE REVIEW

File: azbfs/url_directory.go [hunk 1]
```diff
@@ -130,13 +130,13 @@ func (d DirectoryURL) FileSystemURL() FileSystemURL {
 // list all files/directories inside the directory. Use an empty Marker to start enumeration from the beginning.
 // After getting a segment, process it, and then call ListDirectorySegment again (passing the the previously-returned
 // Marker) to get the next segment.
-func (d DirectoryURL) ListDirectorySegment(ctx context.Context, marker *string, recursive bool) (*DirectoryListResponse, error) {
+func (d DirectoryURL) ListDirectorySegment(ctx context.Context, marker Marker, recursive bool) (*DirectoryListResponse, error) {
```
Comment (shyaler) [inline azbfs/url_directory.go:133]: Why do you have to change it? Is isn't it asbfs sdk?

File: azbfs/zz_response_model.go [hunk 2]
```diff
@@ -2,9 +2,38 @@ package azbfs
...
+type Metadata map[string]string
+
+const mdPrefix = "x-ms-meta-"
+
+const mdPrefixLen = len(mdPrefix)
+
+// NewMetadata returns user-defined key/value pairs.
+func NewMetadata(properties string) (Metadata, error) {
+	md := Metadata{}
+	pairs := strings.Split(properties, ",")
```
Comment (shyaler) [inline azbfs/zz_response_model.go:19]: pairs of what? consider using meaningful name

File: cmd/removeEnumerator.go [hunk 3]
```diff
@@ -313,7 +313,8 @@ func removeSingleBfsResource(ctx context.Context, urlParts azbfs.BfsURLParts, p
 		return "Successfully removed directory: " + urlParts.DirectoryOrFilePath, nil
 	} else {
 		for {
-			listResp, err := directoryURL.ListDirectorySegment(ctx, &marker, recursive)
+			marker := azbfs.Marker{}
```
Comment (shyaler) [inline cmd/removeEnumerator.go:316]: put inline in ListDirectorySegment

File: cmd/sync.go [hunk 4]
```diff
@@ -262,7 +262,7 @@ func (raw *RawSyncCmdArgs) Cook() (cookedSyncCmdArgs, error) {
 	}
 
 	cooked.followSymlinks = raw.FollowSymlinks
-	if cooked.fromTo != common.EFromTo.LocalFile() && cooked.fromTo != common.EFromTo.LocalBlob() { // Follow symlinks in sync is supported only for local -> file or local -> blob
+	if cooked.followSymlinks && cooked.fromTo != common.EFromTo.LocalFile() && cooked.fromTo != common.EFromTo.LocalBlob() { // Follow symlinks in sync is supported only for local -> file or local -> blob
```
Comment (shyaler) [inline cmd/sync.go:265]: wow :0

File: cmd/zc_traverser_blobfs_account.go [hunk 5]
```diff
@@ -105,7 +107,7 @@ func (t *BlobFSAccountTraverser) Traverse(preprocessor objectMorpher, processor
 
 	for _, v := range fsList {
 		fileSystemURL := t.accountURL.NewFileSystemURL(v).URL()
-		fileSystemTraverser := newBlobFSTraverser(&fileSystemURL, t.p, t.ctx, true, t.incrementEnumerationCounter)
+		fileSystemTraverser := newBlobFSTraverser(&fileSystemURL, t.p, t.ctx, true, t.incrementEnumerationCounter, false, false, nil, nil, nil, 0, time.Time{}, common.CFDModeFlags.NotDefined(), false, nil)
```
Comment (shyaler) [inline cmd/zc_traverser_blobfs_account.go:110]: please use named params

File: azbfs/url_directory.go (hunk 1 above)
Comment (OmarKhatib158) [reply azbfs/url_directory.go:133]: We have two sdks:
* the azblob sdk which is public and implemented by storage team and used for blob.
* this (azbfs) sdk which is implemented by Azcopy and used for blobfs.

File: azbfs/zz_response_model.go (hunk 2 above)
Comment (OmarKhatib158) [reply azbfs/zz_response_model.go:19]: key/ value of the metadata. Will rename

File: cmd/removeEnumerator.go (hunk 3 above)
Comment (OmarKhatib158) [reply cmd/removeEnumerator.go:316]: The code is auto formatted in visual code

File: cmd/zc_traverser_blobfs.go [hunk 6]
```diff
@@ -87,156 +132,394 @@ func (t *blobFSTraverser) getFolderProps() (p contentPropsProvider, size int64)
...
-			var entityType common.EntityType
-			lmt := v.LastModifiedTime()
-			if v.IsDirectory == nil || *v.IsDirectory == false {
-				entityType = common.EEntityType.File()
-				contentProps = md5OnlyAdapter{md5: t.getContentMd5(t.ctx, dirUrl, v)}
-				size = *v.ContentLength
-			} else {
-				entityType = common.EEntityType.Folder()
-				contentProps, size = t.getFolderProps()
+	if bfsURLParts.DirectoryOrFilePath != "" {
```
Comment (shmuelberezin) [inline cmd/zc_traverser_blobfs.go:304]: What is the need for the check here?

File: cmd/zc_traverser_blobfs.go [hunk 7]
```diff
@@ -87,156 +132,394 @@ func (t *blobFSTraverser) getFolderProps() (p contentPropsProvider, size int64)
...
+				if t.recursive {
+					// If recursive is turned on, add sub directories to be
+					// in case of target sync, we will not enqueue this dir using enqueue dir, since the enumerateOneDir should enumerates only folders that exists in the source
+					// So we will not enumerate destination directories that was deleted in the source
+					if !targetTraverser {
+						enqueueDir(currentDirURL.NewDirectoryURL(dirInfo))
+					} else {
+						// In case of folder that exist only in the destination, the folder should enumerate by the target traveres
+						// So that the comparere will be able to know that this directory should be treated
+						if err == nil {
```
Comment (shmuelberezin) [inline cmd/zc_traverser_blobfs.go:472]: What is the reason to check for err == nil here? Shouldn't it be handled above after the call to GetProperties?

File: cmd/zc_traverser_blobfs.go [hunk 8]
```diff
@@ -87,156 +132,394 @@ func (t *blobFSTraverser) getFolderProps() (p contentPropsProvider, size int64)
 	return noContentProps, 0
 }
 
-func (t *blobFSTraverser) Traverse(preprocessor objectMorpher, processor objectProcessor, filters []ObjectFilter) (err error) {
-	bfsURLParts := azbfs.NewBfsURLParts(*t.rawURL)
+func (t *blobFSTraverser) writeToErrorChannel(err ErrorFileInfo) {
```
Comment (shmuelberezin) [inline cmd/zc_traverser_blobfs.go:135]: I don't find a call to this func...

File: cmd/zc_traverser_blobfs.go [hunk 9]
```diff
@@ -87,156 +132,394 @@ func (t *blobFSTraverser) getFolderProps() (p contentPropsProvider, size int64)
...
+				ContainerName:     bfsURLParts.FileSystemName,
+				isFolderEndMarker: true,
+				isFinalizeAll:     FinalizeAll,
+			}
+			enqueueOutput(storedObject, nil)
 		}
+		return nil
+	}
+
+	if targetTraverser {
```
Comment (shmuelberezin) [inline cmd/zc_traverser_blobfs.go:521]: As far as I see, you handle only traversing the 'isTargetSync' case.
This is enough for the mover agent use case but it's different from the way the blob and file traversers are implemented.
I'm not sure there is a need to invest in adding the support for these scenarios. Let's discuss with @shyaler

File: cmd/zc_traverser_blobfs.go (hunk 6 above)
Comment (OmarKhatib158) [reply cmd/zc_traverser_blobfs.go:304]: If we have no sub-path then we don't need to check if it is file or directory, we already know it is the container.

File: cmd/zc_traverser_blobfs_account.go (hunk 5 above)
Comment (OmarKhatib158) [reply cmd/zc_traverser_blobfs_account.go:110]: Not supported for functions arguments in go :(

File: cmd/zc_traverser_blobfs.go (hunk 7 above)
Comment (OmarKhatib158) [reply cmd/zc_traverser_blobfs.go:472]: Removed

File: cmd/zc_traverser_blobfs.go (hunk 8 above)
Comment (OmarKhatib158) [reply cmd/zc_traverser_blobfs.go:135]: Added

**Review by shmuelberezin — APPROVED** [review]

**Pull request merged**